components/ai_chat/core/browser/ai_chat_service.cc

```diff
@@ -0,0 +1,373 @@
+    conversation_observations_.RemoveObservation(conversation_handler);
+    conversation_handlers_.erase(uuid);
+    conversations_.erase(uuid);
+    DVLOG(1) << "Erased conversation (" << uuid << "). Now have "
+             << conversations_.size() << " Conversation metadata items and "
+             << conversation_handlers_.size()
+             << " ConversationHandler instances.";
+    OnConversationListChanged();
+  }
+}
+
+void AIChatService::OnConversationEntriesChanged(
+    ConversationHandler* handler,
+    std::vector<mojom::ConversationTurnPtr> entries) {
+  auto conversation_it = conversations_.find(handler->get_conversation_uuid());
+  CHECK(conversation_it != conversations_.end());
+  auto& conversation = conversation_it->second;
+  if (!conversation->has_content && !entries.empty()) {
+    // This conversation is now visible
+    conversation->has_content = true;
+    OnConversationListChanged();
+    if (ai_chat_metrics_ != nullptr) {
+      if (entries.size() == 1) {
+        ai_chat_metrics_->RecordNewChat();
+      }
+      if (entries.back()->character_type == mojom::CharacterType::HUMAN) {
+        ai_chat_metrics_->RecordNewPrompt();
+      }
+    }
+  }
+  // TODO(petemill): Persist the entries, but consider receiving finer grained
+  // entry update events.
+}
+
+void AIChatService::OnClientConnectionChanged(ConversationHandler* handler) {
+  MaybeEraseConversation(handler);
+}
+
+void AIChatService::GetVisibleConversations(
+    GetVisibleConversationsCallback callback) {
+  std::vector<mojom::ConversationPtr> conversations;
+  for (const auto& conversation : FilterVisibleConversations()) {
+    conversations.push_back(conversation->Clone());
+  }
+  std::move(callback).Run(std::move(conversations));
+}
+
+void AIChatService::BindConversation(
+    const std::string& uuid,
+    mojo::PendingReceiver<mojom::ConversationHandler> receiver,
+    mojo::PendingRemote<mojom::ConversationUI> conversation_ui_handler) {
+  ConversationHandler* conversation = GetConversation(uuid);
+  CHECK(conversation) << "Asked to bind a conversation which doesn't exist";
+  conversation->Bind(std::move(receiver), std::move(conversation_ui_handler));
+}
+
+void AIChatService::BindObserver(
+    mojo::PendingRemote<mojom::ServiceObserver> observer) {
+  observer_remotes_.Add(std::move(observer));
+}
+
+bool AIChatService::HasUserOptedIn() {
+  return ai_chat::HasUserOptedIn(profile_prefs_);
+}
+
+bool AIChatService::IsPremiumStatus() {
+  return ai_chat::IsPremiumStatus(last_premium_status_);
+}
+
+std::unique_ptr<EngineConsumer> AIChatService::GetDefaultAIEngine() {
+  return model_service_->GetEngineForModel(model_service_->GetDefaultModelKey(),
+                                           url_loader_factory_,
+                                           credential_manager_.get());
+}
+
+size_t AIChatService::GetInMemoryConversationCountForTesting() {
+  CHECK(conversations_.size() == conversation_handlers_.size());
```

Comment:
Well yes, without any DB then they should always be the same size, so this was me putting in a verification for that for the time-being during testing. But since this is only concerned with active conversations then I've changed it to return `conversation_handlers_.size()` and removed the CHECK - thanks.